packages/vite/src/node/server/moduleGraph.ts

```diff
@@ -150,15 +150,24 @@ export class ModuleGraph {
     const prevImports = mod.importedModules
     const nextImports = (mod.importedModules = new Set())
     let noLongerImported: Set<ModuleNode> | undefined
+    let resolvePromises = []
     // update import graph
     for (const imported of importedModules) {
-      const dep =
-        typeof imported === 'string'
-          ? await this.ensureEntryFromUrl(imported, ssr)
-          : imported
-      dep.importers.add(mod)
-      nextImports.add(dep)
+      if (typeof imported === 'string') {
+        resolvePromises.push(
+          this.ensureEntryFromUrl(imported, ssr).then((dep) => {
+            dep.importers.add(mod)
+            nextImports.add(dep)
+          }),
+        )
+      } else {
+        imported.importers.add(mod)
+        nextImports.add(imported)
+      }
     }
+
+    resolvePromises.length && (await Promise.all(resolvePromises))
```

Comment:
Nice find! I wonder if we can defer and only await `Promise.all` at the bottom? IIUC `ensureEntryFromUrl` is isolated work in the module graph, unless the two urls passed could point to the same file/id, but I think that's rare and wouldn't cause any issues.

Reply:
Do you mean await `resolvePromises` that includes (`importedModules` and `acceptedHmrDeps`)'s `ensureEntryFromUrl`  only once at the bottom of `updateModuleInfo`?

Reply:
Yeah I mean waiting for all the modules at once. Not sure if we can make this non-blocking, but I think  it should be up to the consumer of `updateModuleInfo` to await or not.

Reply:
Emm, it seems `acceptedHmrDeps. ensureEntryFromUrl ` can't run in parrallel with   `importedModules. ensureEntryFromUrl `.

Reply:
Thanks for checking it! It might be related to:

https://github.com/vitejs/vite/blob/0f9ad6896ae93f79cc7ef6fbb2fd56047d719245/packages/vite/src/node/server/moduleGraph.ts#L165

Let's ignore this for now then 👍